### DISCUSSION TITLE
[测试版] 3.6.0a，新增 Cloudflare Workers 服务端

### DISCUSSION THREAD

**Discussion by SeaHOH** [SeaHOH/GotoX#142] [公告]
https://github.com/SeaHOH/GotoX/releases/tag/v3.6.0a

没有部署过 GAE 服务端的新用户可以直接使用，部署方法可查看 wiki。但建议老用户暂时仍然使用 GAE，直到 CFWorkers 服务端更加完善易用。

同时，欢迎反馈错误和建议。

**测试阶段会随时调整代码，反馈前请先看看你是否需要更新。**

现已发现问题，勾选代表已解决：
- [x] 无法上传数据。
- [x] 无法转发 WebSocket。
- [x] 大图片下载过慢。
- [x] [混淆的邮件地址](https://support.cloudflare.com/hc/en-us/articles/200170016-What-is-Email-Address-Obfuscation-)无法自动解码。
- [x] 无法正确处理 Range 请求返回的响应。
- [ ] 不支持非标准端口（**存疑，请求更多测试样本**）。
- [ ] 标准不支持 chunked 上传，`fetch()` 要读取全部数据才开始发出请求，对大量数据是个障碍，造成多余的延时和内存使用，单个 Worker 内存使用上限为 128 MB。

**Comment by standritsu** [comment]
纯CFW模式的话8086还是要走GAE

**Comment by SeaHOH** [comment]
现在没有纯 CFW 模式，等完善之后，如果测试性能足够的话会增加此种模式。

**Comment by lshilshi** [comment]
需要有域名？还是说可以自由输入？

**Comment by SeaHOH** [comment]
Cloudflare  为 Workers 提供有免费域名，参考 wiki 部署即可。

**Comment by hg123** [comment]
youtube视频看一会后就会出现错误

**Comment by k79e** [comment]
https://developers.cloudflare.com/workers/reference/apis/streams/
这里有几个chunk词出现 还提到buffer

> read: A promise that returns the next available chunk of data being passed through the reader queue.

> read(buffer): Returns a promise with the next available chunk of data read into a passed-in buffer.

https://blog.cloudflare.com/edge-side-includes-with-cloudflare-workers/
In this portion, we’re **receiving the chunks** and searching for ESI blocks and if found, getting the dynamic parts. You can see that when the regex is matching, we’re calling a subRequest() function. This function will fetch the content of the ESI destination and return + encode the received payload.

**Comment by k79e** [comment]
#143 那里我发现png变渐进的了. 是不是什么压缩规则出了问题呢. 
应该只压缩文本. 资源文件等排除才对.

//我直接下的master版

**Comment by SeaHOH** [comment]
上面提到的是 `fetch()` 无法实现用 chunked 方式发起请求，其结果 `Response` 和最终返回都能正常支持。 https://fetch.spec.whatwg.org/#http-network-fetch 这个解释有点谜，为什么 HTTP/2 不能用？`fetch()` 发起于内部网络，肯定不是 HTTP/1.X。

关于图片，你可以查看 header，应该没有压缩，只是大文件下载速度太慢。`fetch()` 有几个优化开关，其中图片可以被优化，但是这样只能得到压缩过的，以后可以支持设置部分开关，感兴趣的可以查看 worker 源码 `fetch()` 的 `cf` 选项并提出建议，我把能用的都写进去了，具体解释参考 https://developers.cloudflare.com/workers/reference/apis/request/ 。

另，暂时不建议用来看 U2B 等视频网站，大文件优化属于收尾部分的工作。

**Comment by k79e** [comment]
图片我就是看他非渐进的图片变成渐进了. 载入的时候很明显是变成模糊的. 而且速度很快 然后慢慢变清晰. 我也不知道为啥. 正常情况下是从上到下载入的. 用了代理才那样....

> Responses over an HTTP/2 connection will always have the empty byte sequence **as status message** as HTTP/2 does not support them.

他不是说的支持h2么 如果不支持怎么会有**通过h2**的响应.
h2不支持的应该是**一种**状态信息. 所以用空字节代替.

**Comment by k79e** [comment]
关于图片 很诡异 看起来是cf处理过的. 因为源文件不属于cf网络 没有cdn.

Last-Modified: Sun, 0
Cf-Cache-Status: MISS

第二次载入之后发现多了头部
Last-Modified: Sun,
Cf-Cache-Status: HIT
Cf-Bgj: **imgq**:100
Cf-Polished: **origSize**=885498
.....
Content-Length: **778518**

所以优化图片的一个参数好像默认开启了. 能关闭的话建议默认关闭. 或者来个开关能控制也好.

//发现开关了 当前开关是无损. 我改下去. 见鬼了配置off也关不掉 不知道为啥 这跟浏览器缓存有关系么 
过段时间我再看看去.....

图片压缩bug大了 昨天测试大图发现就小的有渐进效果 其他都没什么变化.(原来是cf会自动忽略)

现在我发现有效果了 原来是cf把资源给缓存了 内容更新有点延时

原来是缓存里面已经有无损压缩的资源 压缩还关闭的话 他还是会返回更小的那个资源.

**Comment by k79e** [comment]
在这个页面里面提到了 写chunk 不知道是不是发起请求/上传的意思
https://developers.cloudflare.com/workers/reference/apis/streams/
write(chunk): **Writes a chunk of data** to the writer and returns a promise to indicate if the operation succeeded or failed.

https://developer.mozilla.org/zh-CN/docs/Web/API/Streams_API/%E4%BD%BF%E7%94%A8%E5%8F%AF%E8%AF%BB%E6%96%87%E4%BB%B6%E6%B5%81
Creating your own custom readable stream
这里提到构造一个自定义可读的流 不知道是不是发送数据.

这数据post出去他不会自动分块么?

https://javascript.info/fetch
这地方提到post二进制数据
POST requests
  Blob/BufferSource to send binary data,

https://github.com/whatwg/fetch/issues/966
好像他们说那个功能还没做好......

**Comment by k79e** [comment]
我下载一个文件 然后停止 重复几次程序卡了
不知道这是不是bug

![2020-02-20_201503](https://user-images.githubusercontent.com/6483602/74933089-3ef1af80-541e-11ea-90bc-3005dbc9ae05.png)

**Comment by SeaHOH** [comment]
已经找到转发 websocket 的具体条件，等我再完善下细节就更新。

图片的事情先不着急，后面我再专门测试，你也可以试着自己修改 CF 开关选项。

`fetch()` 要完全读取 `body` 负载才开始发送请求，这个实在是个大问题。

> 我下载一个文件 然后停止 重复几次程序卡了
> 不知道这是不是bug

这确实算 bug，我已经打算转移到原生 async 框架，到时只支持 py37+。

**Comment by k79e** [comment]
图片没问题了 我改了代码里面的设置. 改off就没事了.

**Comment by lshilshi** [comment]
@SeaHOH config.ini中 [cfw] 段的password设置为自己worker的登录密码？

**Comment by k79e** [comment]
js代码里面的密码

**Comment by SeaHOH** [comment]
password 为使用密码，并非 Worker 登录密码，需要修改服务端 js 来自行设置。

**Comment by lshilshi** [comment]
> password 为使用密码，并非 Worker 登录密码，需要修改服务端 js 来自行设置。

那如果是按你wiki里面的复制
password为空？

**Comment by SeaHOH** [comment]
你部署时就已经查看过 index.js 的内容，这是没有使用密码的，我也不可能设置个默认密码，设置它需要自行修改 js 内容。

**Comment by ShadowEmbrace** [comment]
隔壁过来的，谢谢大佬。不知道我这个反馈有没有价值。
海南电信，部署了worker，没输入gaeid，建了he的隧道，导入了xxnet的ipv6。目前速度很慢，不过好的地方是xxnet仅用ipv6基本是打不开，gotox等一等还是可以打开google.com。是不是开着运行一段时间会好转？

**Comment by SeaHOH** [comment]
> 是不是开着运行一段时间会好转？

GotoX 对谷歌 IP 进行了分类，可以连接一般服务的 IP 远多于能够连接 GAE 的 IP，使用这些 IP 就能不用通过代理直接访问谷歌。由于近几天干扰很严重，需要运行一段时间才能筛选出相对更稳定些的 IP。就谷歌一般服务而言，IPv4 的稳定 IP 更多些。

**Comment by hg123** [comment]
有时候会出现如下提示
```
Exception happened during processing of request from ('127.0.0.1', 61604)
Traceback (most recent call last):
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyServer.py", line 125, in process_request_thread
    self.finish_request(request, client_address)
  File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\socketserver.py", line 360, in finish_request
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 129, in __init__
    self.handle()
  File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\server.py", line 426, in handle
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 186, in handle_one_request
    self.do_CONNECT()
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 288, in do_CONNECT
    self.do_action()
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 232, in do_action
    getattr(self, self.action)()
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 1117, in do_FAKECERT
    self.handle()
  File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\server.py", line 426, in handle
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 188, in handle_one_request
    self.do_METHOD()
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 314, in do_METHOD
    self.do_action()
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 232, in do_action
    getattr(self, self.action)()
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 665, in do_CFW
    raise e
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 650, in do_CFW
    response, data, need_chunked, ws_ok = self.handle_response_headers(response)
  File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 445, in handle_response_headers
    response_headers = {k.title(): v for k, v in response.headers.items()
AttributeError: 'NoneType' object has no attribute 'headers'
```

**Comment by SeaHOH** [comment]
@hg123 你是否更新到了 master 版本？

**Comment by hg123** [comment]
没更新到master版

**Comment by k79e** [comment]
ws是不是还不支持 
我用curl测试了下发现报错500
HTTP/1.1 500 Internal Server Error
我用的3.6.0a 是不是版本老了

2020-03-15 22:04:05
果然是老了 我用的比较早的master版 更新之后没事了

**Comment by hg123** [comment]
观察了一下，好像是在连接不上服务器的时候就会出现我说的这种错误，一般在客户端重启之后又可以使用了。

**Comment by SeaHOH** [comment]
> 没更新到master版

更新没？

**Comment by hg123** [comment]
我现在服务端和客户端都更新了，我再观察一下。

**Comment by hg123** [comment]
刚刚又更新到最新版 打不开了 闪一下就退出了

**Comment by k79e** [comment]
好像跟有个什么东西要下载有关...
//我记得有次修改提到过把什么组建给剔除了 改为在线下载了
我前几天也是 打开报错 可是开了几次发现又正常了

**Comment by hg123** [comment]
我是打开了多次啊 一样闪退 我再试试吧

》》》刚才又试了一下 还是闪退

**Comment by Gravites** [comment]
> 我是打开了多次啊一样闪退我再试试吧
> 
> 》》》刚才又试了一下还是闪退
一样闪退

**Comment by SeaHOH** [comment]
使用 Windows 版本的用户，请手动覆盖更新时，请删除下面这个文件：
**\python\site-packages\helpers_win32.egg**

仅此一次，以后应该不会改了。

**Comment by k79e** [comment]
奇怪了 我也有这个文件 但是可以正常用.....
我也不知道上次报错咋样恢复正常了 就是开了几次都报错 然后关了没管它
后来打开又能用了.....

**Comment by SeaHOH** [comment]
使用 Windows 托盘工具才会出现更新后无法启动，其它没问题。
我没办法删除你们的旧文件，只能你们自己动手。

**Comment by hg123** [comment]
> 
> 
> 有时候会出现如下提示
> 
> ```
> Exception happened during processing of request from ('127.0.0.1', 61604)
> Traceback (most recent call last):
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyServer.py", line 125, in process_request_thread
>     self.finish_request(request, client_address)
>   File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\socketserver.py", line 360, in finish_request
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 129, in __init__
>     self.handle()
>   File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\server.py", line 426, in handle
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 186, in handle_one_request
>     self.do_CONNECT()
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 288, in do_CONNECT
>     self.do_action()
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 232, in do_action
>     getattr(self, self.action)()
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 1117, in do_FAKECERT
>     self.handle()
>   File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\server.py", line 426, in handle
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 188, in handle_one_request
>     self.do_METHOD()
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 314, in do_METHOD
>     self.do_action()
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 232, in do_action
>     getattr(self, self.action)()
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 665, in do_CFW
>     raise e
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 650, in do_CFW
>     response, data, need_chunked, ws_ok = self.handle_response_headers(response)
>   File "C:\Users\Administrator\Desktop\GotoX-3.6.0a-cp37-win32\local\ProxyHandler.py", line 445, in handle_response_headers
>     response_headers = {k.title(): v for k, v in response.headers.items()
> AttributeError: 'NoneType' object has no attribute 'headers'
> ```
```
Exception happened during processing of request from ('127.0.0.1', 58086)
Traceback (most recent call last):
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyServer.py", line 125, in process_request_thread
    self.finish_request(request, client_address)
  File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\socketserver.py", line 360, in finish_request
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 129, in __init__
    self.handle()
  File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\server.py", line 426, in handle
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 186, in handle_one_request
    self.do_CONNECT()
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 288, in do_CONNECT
    self.do_action()
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 232, in do_action
    getattr(self, self.action)()
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 1119, in do_FAKECERT
    self.handle()
  File "D:\obj\windows-release\37win32_Release\msi_python\zip_win32\server.py", line 426, in handle
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 188, in handle_one_request
    self.do_METHOD()
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 314, in do_METHOD
    self.do_action()
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 232, in do_action
    getattr(self, self.action)()
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 667, in do_CFW
    raise e
  File "C:\Users\Administrator\Desktop\GotoX\local\ProxyHandler.py", line 649, in do_CFW
    response = cfw_fetch(self.command, self.url, request_headers, payload)
  File "C:\Users\Administrator\Desktop\GotoX\local\CFWFetch.py", line 93, in cfw_fetch
    realurl=realurl)
  File "C:\Users\Administrator\Desktop\GotoX\local\HTTPUtil.py", line 921, in request
    elif len(dns[hostname]) == 1 and timeout < 5:
TypeError: object of type 'object' has no len()
```
最新版 我这里还是出现这种错误

**Comment by SeaHOH** [comment]
已修复。不过这个错误和之前那个不同，之前的早已修复。

**Comment by Destineda** [comment]
> 已修复。不过这个错误和之前那个不同，之前的早已修复。

下载不了 master ， 404

**Comment by SeaHOH** [comment]
> 下载 master 现在404

如果没法直接下载，请使用代理。

**Comment by lxf1009** [comment]
有时候打开解析不出IP，直接点重启的话还出错，只能关闭重新打开好几次才解析出IP，请问有什么解决办法

**Comment by k79e** [comment]
是不是dns出现黄字的现象呢? 这个我当初就发现了......

**Comment by SeaHOH** [comment]
@lxf1009 不要关闭 overhttps，然后把 overhttps 优先级设置为最高。

鉴于 overhttps 还算稳定，反而用户的其它设置和环境不确定，以后会把这个优先级作为默认设置。

**Comment by lxf1009** [comment]
overhttps排在最前面可以了，几乎每次都成功。

**Comment by standritsu** [comment]
05:04:42 E L6:54503-> 无法解析主机：'www.biketo.com'，路径：'/'，请检查是否输入正确！
05:04:43 E L6:54509-> 无法解析主机：'www.biketo.com'，路径：'/favicon.ico'，请检查是否输入正确！

我会出现红字的这种错误，重启一下GOTOX就好了，域名不一定，大概率是.cn或者是国内的网站

**Comment by k79e** [comment]
我上次就提到过这个问题 只不过我是本地回环的dns(我估计用其他dns也一个样吧)
你看我.1的ip也不是一层套一层的 我不用.2的用.1他肯定还出问题...
反正当初出问题就是秒报错 流量监控都表示程序根本没发出流量.....
对了这个应该能给linux用 那样测试比较方便....
不对啊 我发现的是黄字 楼上的应该是红字.... 不是一个东西

2020-03-23 14:30:55
奇怪 今天随便查了几下 发现没黄字了.....

**Comment by hg123** [comment]
最新版会提示：CFW 'xxxx.workers.dev' 工作异常：530，有时候会提示： 工作异常：404 Not Found。
必须多次刷新网页， 网页才能显示。

**Comment by SeaHOH** [comment]
对，打印出来以便收集处理。现已处理错误状态 429，代表一分钟内请求超过 1000 次，会暂停请求 30 秒。以后会添加更多错误处理。

话说 530 后面没有文字描述吗？

**Comment by hg123** [comment]
530后面什么字都没有 一般是下面这样
 CFW xxxxxx.workers.dev' 工作异常：530
xx.xx.xx.xx:xxxxx->xx.xx.xx.xx "CFW POST https://api.twitter.com/1.1/jot/client_event.json HTTP/1.1" 530
好像是这个ip无法连接还是怎么样才出现这种错误

看了下日志 出现530的是某个IP 出现404的是另一个IP

**Comment by SeaHOH** [comment]
同一个域名的请求会在短时间内重复触发 530 状态吗？

404 的 IP 麻烦收集一下贴出来，我需要测试它们。

**Comment by hg123** [comment]
404的ip目前碰到的就只有104.17.175.182和104.17.175.33
触发530状态的网址不一样 但都与twitter有关 因为我访问最多的网站就是twitter ip是104.17.175.55
刚才又看到这个提示 工作异常：302 Found ip是 104.17.176.11

**Comment by hg123** [comment]
又有个404IP 104.17.176.40

**Comment by k79e** [comment]
返回 HTTP 错误 530 时会同时显示 1XXX 错误。在 Cloudflare 帮助中心中搜索具体的 1XXX 错误，以了解故障排除信息。

**Comment by SeaHOH** [comment]
df377f8ac58fffe8b669ed4d17169e6e55b8481f 已处理。后续如果还有，请反馈。谢谢啦~

**Comment by standritsu** [comment]
06:03:42 W CFW '********.workers.dev' 工作异常：500 Internal Server Error
06:03:43 W L6:56372->[2606:4700::6811:b0d2] "CFW POST https://syndication.twitter.com/i/jot HTTP/1.1" 500 3741
06:03:43 I L6:56372->[2606:4700::6811:b0d2] "CFW GET https://syndication.twitter.com/cdn-cgi/styles/cf.errors.css HTTP/1.1" 200 -

刷到这个

**Comment by SeaHOH** [comment]
经测试，IP 本身没有问题，这个 500 状态属于正常响应，已移除警告。

**Comment by Quantum2Bits** [comment]
用CFW 不能打开bbc.com。今天是了很久都不行。用GAE可以。GAE现在很慢。

**Comment by SeaHOH** [comment]
发送空白 SNI 或者直接访问 `www.bbc.com`（默认规则此域名已配置空白 SNI）。

**Comment by hg123** [comment]
最近时不时地会连不上服务端 
 104.17.175.78 _create_ssl_connection 'xxx.xxx.workers.dev' 返回 timeout('timed out')，重试
 104.17.176.17 _create_ssl_connection 'xxx.xxx.workers.dev' 返回 timeout('timed out')，重试
这时域名网址无法直连却可以通过其他代理访问 过一段时间后又能恢复正常
是我的网络原因吗？

这段时间基本上连不上服务端了 如果长时间不用 可以短暂连一会 然后很快就断开了
另外好像只要与cloudflare有关的其他网站也都打不开了

**Comment by SeaHOH** [comment]
我这里没发现此类现象，可能你的网络不稳定，同时发起的请求是否特别多，其它国外服务能否正常访问？

**Comment by hg123** [comment]
我之前都是浏览同样的网页却没有出现这个问题 ，近一个月来却极不稳定 。
连不上服务器后，好像国内使用cloudflare的cdn的网站也打不开了
你说的其它国外服务器是指没有被墙的吗？

应该是我网络的问题 ，家里的宽带只是偶尔会出现这种情况 ，但很快就会恢复。
但办公室的电脑只要一连被墙网站，很快就会被防火墙之类的切断连接似的

**Comment by hg123** [comment]
刚刚把通过cfw代理的网站全部换到gae代理，速度极快。前段时间一直很慢。

**Comment by SeaHOH** [comment]
> 你说的其它国外服务器是指没有被墙的吗？

是的。

> 刚刚把通过cfw代理的网站全部换到gae代理，速度极快。

谷歌经常折腾，反正免费的越来越不好用了。用的人少了，线路丢包情况可能会比之前好。

**Comment by hg123** [comment]
刚才试了下 如果用了cfw被阻断后 国外没有被墙的网站全都上不了

如果用gae的话就不会出现这种情况

**Comment by SeaHOH** [comment]
明显是你的网络问题。以后会尝试增加更多的远程代理，暂时就这样吧。

**Comment by k79e** [comment]
现在看起来支持chunk了?
https://streams.spec.whatwg.org/#read-request-chunk-steps

**Comment by SeaHOH** [comment]
具体没看太明白，反正要服务器添加支持才能使用，有没有具体实现还是问号。
想来也不可能很快大规模普及，尤其考虑到安全和兼容问题，大部分没有实际需求的服务商都不愿意部署吧。

**Comment by k79e** [comment]
> Chunks are enqueued into the stream by the stream’s underlying source. They can then be read one at a time via the stream’s > public interface, in particular by using a readable stream reader acquired using the stream’s getReader() method.

getReader()  https://streams.spec.whatwg.org/#rs-get-reader
有样例的样子



还有可写数据流
**Chunks are written to the stream** via its public interface, and are passed one at a time to the stream’s underlying sink. For web developer-created streams, the implementation details of the sink are provided by an object with certain methods that is passed to the WritableStream() constructor.

看样子最新版是能读能写chunk的样子
参考
2.2. Writable streams
5.2.3. The underlying sink API
5.2.4. Constructor, methods, and properties

**Comment by k79e** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/Streams_API

https://developer.mozilla.org/zh-CN/docs/Web/API/ReadableStream
https://developer.mozilla.org/zh-CN/docs/Web/API/WritableStream

**Comment by k79e** [comment]
就关于你提到的非标准端口 我前段时间测了 用的反代 可以连上 (我随便填了好几个端口 都连上了 应该没啥问题)
//但是不可以回接国内 这个影响不大
所以worker是支持那个的 cdn就不行了 只有那么几个端口可以用

**Comment by SeaHOH** [comment]
好像只是接口变化，但 Request 本质没变，还是要完全接收数据才开始。

**Comment by SeaHOH** [comment]
不支持非标准端口，是指被代理的请求端口，不是指连接 worker 服务器使用的端口。

**Comment by k79e** [comment]
就是目标地址 可以是其他端口的
worker不是只有web端口 怎么能随便输入端口呢....
